### PR TITLE
#242 coline railmap

### LINKED ISSUE #41: Hope to add co-line instructions for Shanghai Metro style (like the co-line ends of lines 3 and 4)
Since Shanghai Metro Lines 3 and 4 are co-line ends, it means that the location of the co-line segment can be shown, and it can also become more effective. Here are some suggestions
1. When adding a common line segment representation to a line, it means that another line indication needs to be added, which may increase the algorithm complexity, so the indication of another line may not be needed initially, but if the file input port of another line can be added , Meaning the algorithm will be simpler (recommended)
2. Other cities have not made a collinear end statement, please also pay attention to the official website of the subway company in other cities, if there is any, you can add it as soon as possible

### PR DESCRIPTION
**Pull request by thekingofcity**
Let's fix #41 one by one with railmap first.

The coline implementation deserves more explanation as the development process goes.

**TL;DR** This is a Style(mtr/gzmtr/**shmetro**)-specific and Canvas(param/ui/**svg**)-specific PR. (of course coline related UI, param, and actions are added, but they are added in a non-breaking and non-changing existing features way)

```
export interface ColineInfo {
    from: string;
    to: string;
    colors: InterchangeInfo[];
    display: boolean;
}


/**
 * Dictionary of configuration parameters for RMG, stored in `localStorage` as string.
 */
export interface RMGParam {
    // all existing params
    coline: ColineInfo[];
}
```

So basically coline adds its *coline* segments in its place in `param` and when the SVGs need to be drawn, coline's *line* and *stations* are calculated on the fly. This mainly includes line `path` from which station to which station and its colors.

UI and redux things are responsible for the coline fields and have nothing to do with the existing param and algorithms so this is a non-breaking change as an empty coline field won't break anything as others won't check it at all.

### CODE REVIEW

**Comment by thekingofcity** [comment]
Looks like there are some glitches in the unit tests. Would you like to fix them or force merge? By the way the PR need to be merged instead of squashed.
